include/vcpkg/base/diagnostics.h

```diff
@@ -0,0 +1,261 @@
+#pragma once
+
+#include <vcpkg/base/expected.h>
+#include <vcpkg/base/messages.h>
+#include <vcpkg/base/optional.h>
+
+#include <mutex>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace vcpkg
+{
+    enum class DiagKind
+    {
+        None,    // foo.h: localized
+        Message, // foo.h: message: localized
+        Error,   // foo.h: error: localized
+        Warning, // foo.h: warning: localized
+        Note,    // foo.h: note: localized
+        COUNT
+    };
+
+    struct TextRowCol
+    {
+        // '0' indicates uninitialized; '1' is the first row/column
+        int row = 0;
+        int column = 0;
+    };
+
+    struct DiagnosticLine
+    {
+        template<class MessageLike, std::enable_if_t<std::is_convertible_v<MessageLike, LocalizedString>, int> = 0>
+        DiagnosticLine(DiagKind kind, MessageLike&& message)
+            : m_kind(kind), m_origin(), m_position(), m_message(std::forward<MessageLike>(message))
+        {
+        }
+
+        template<class MessageLike, std::enable_if_t<std::is_convertible_v<MessageLike, LocalizedString>, int> = 0>
+        DiagnosticLine(DiagKind kind, StringView origin, MessageLike&& message)
+            : m_kind(kind), m_origin(origin.to_string()), m_position(), m_message(std::forward<MessageLike>(message))
+        {
+        }
+
+        template<class MessageLike, std::enable_if_t<std::is_convertible_v<MessageLike, LocalizedString>, int> = 0>
+        DiagnosticLine(DiagKind kind, StringView origin, TextRowCol position, MessageLike&& message)
```

Comment:
>This design requires origin and position to be separately plumbed through every function call layer down to where the DiagnosticLine is created -- typically very close to the leaf function that has the appropriate context. I'm afraid many, many functions will end up looking like the following even though they themselves never create a DiagnosticLine -- they just want to call other functions!
>[...]
>`/* Now I can start talking about _what I actually care about_ */`

I argue that for functions that are talking about file / position information like this, position information is not 'something they don't care about', it's essential in implementing what that function does.

I can see an argument for a type that contains:

1. A `DiagnosticContext&`
2. The origin
3. Tracks a cursor of where the currently being parsed position is

and today the type trying to do that is `ParserBase`. I'm not a fan of how `ParserBase` wants and encourages deriving from and making things members when I think a struct containing these 3 members being passed around is a more composable design. I didn't go all the way there here though because

1. That starts to approach the design where callers are expected to provide additional wrapping context I asked for but you have said over time IRL that you do not like, and
2. That would require making this PR bigger since it would change more about how ParserBase is used.

Reply:
Seeing that `ParserBase` tracks state incorrectly since it accepted a column but set `m_start_of_line(m_it)` (so the only valid col was 0 or 1) I've become convinced that 2+3 should go together but 1 should not.